src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+        When `shape` is an integer, this corresponds to
+        the diag(A) where A is (n,n).
+
+        :param shape: Shape of matrix.
+            When integer, corresponds to len(diag(A)).  Otherwise,
+            when a tuple (..., n), the last dimension `n` defines the
+            length of diagonal, while prior dimensions define any
+            stack axes.
+        :param dtype: Datatype, defaults to np.float32.
+        :return: `DiagMatrix` instance.
+        """
+
+        return DiagMatrix(np.zeros(shape, dtype=dtype))
+
+    @staticmethod
+    def ones(shape, dtype=np.float32):
+        """
+        Instantiate ones intialized `DiagMatrix`.
+        When `shape` is an integer, this corresponds to
+        the diag(A) where A is (n,n).
+
+        :param shape: Shape of matrix.
+            When integer, corresponds to len(diag(A)).  Otherwise,
+            when a tuple (..., n), the last dimension `n` defines the
+            length of diagonal, while prior dimensions define any
+            stack axes.
+        :param dtype: Datatype, defaults to np.float32.
+        :return: `DiagMatrix` instance.
+        """
+
+        return DiagMatrix(np.ones(shape, dtype=dtype))
+
+    @staticmethod
+    def eye(shape, dtype=np.float32):
+        """
+        Build a `DiagMatrix` eye (identity) matrix.
+        This is simply an alias for `ones`.
+
+        :param shape: Shape of matrix.
+            When integer, corresponds to len(diag(A)).  Otherwise,
+            when a tuple (..., n), the last dimension `n` defines the
+            length of diagonal, while prior dimensions define any
+            stack axes.
+        :param dtype: Datatype, defaults to np.float32.
+        :return: `DiagMatrix` instance.
+        """
+
+        return DiagMatrix.ones(shape, dtype=dtype)
+
+    def as_blk_diag(self, partition):
+        """
+        Express `DiagMatrix` as a `BlkDiagMatrix` using `partition`.
+
+        :param partition: BlkDiagMatrix partition.
+        :return: `BlkDiagMatrix`
+        """
+        if self.stack_shape != ():
+            raise RuntimeError(
+                f"as_blk_diag only implemented for singletons at this time, received {self.stack_shape}."
+            )
+
+        B = BlkDiagMatrix(partition, dtype=self.dtype)
+        ind = 0
+        for b, p in enumerate(partition):
+            if p[0] != p[1]:
+                raise RuntimeError(
+                    f"Block {b} of partition is not square, {p}."
+                    "Currently DiagMatrix.as_blk_diag only supports square blocks."
+                )
+
+            j = p[0]
+            B[b] = np.diag(self._data[ind : ind + j])
+            ind += j
+
+        return B
+
+    def lr_scale(self, weights=None):
+        """
+        Performs L and R scaling of `weights` by this `DiagMatrix`
+        instance simultaneously, returning a `DiagMatrix`.
+
+        Given `weights` this computes B = A * w * A.T via B = w * A**2
```

Comment:
I don't understand. Why should this hold? Is `W` a matrix here or a vector?

Reply:
w is a vector. In this case they are all conceptually diags although the weights come in as a numpy array (can be changed to diag/agnostic)  Probably those should be `@` signs in the comment, my bad for mixing up notation.  The usual cov decomp transpose is no-op since the diag is sym. Similarly since its a diag they can be commuted around.

Although this operation doesn't need it's own method, this operation in conjunction with the direct (div) `solve` are the two main contributions for the cov2d using FLE... so I gave it it's own method to sort of highlight it.  If you would rather I can just code using the `DiagMatrix` operations where it is relavent, I can totally do that. Just lmk.

```
>>> A
array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 3, 0],
       [0, 0, 0, 4]])
>>> w
array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 4, 0],
       [0, 0, 0, 6]])
>>> A.T @ w @ A
array([[ 1,  0,  0,  0],
       [ 0,  8,  0,  0],
       [ 0,  0, 36,  0],
       [ 0,  0,  0, 96]])
>>> A**2 @ w
array([[ 1,  0,  0,  0],
       [ 0,  8,  0,  0],
       [ 0,  0, 36,  0],
       [ 0,  0,  0, 96]])
```

Reply:
But if `w` is a vector and those are matmuls, `w * A.T` doesn't make any sense. If `w` is n×1 and `A` is n×n`, that multiplication isn't going to work out. Or are you saying that `w` is a diagonal matrix? Then sure, but I'm not sure we need a method just to do this.

Reply:
No problem, just removed it and the associated tests.

Reply:
Cool. Thanks!